src/Extend/Model.php

```diff
@@ -168,15 +170,21 @@ public function hasMany(string $name, string $related, string $foreignKey = null
      *
      * @return self
      */
-    public function relationship(string $name, callable $callable)
+    public function relationship(string $name, $callback)
     {
-        Arr::set(AbstractModel::$customRelations, "$this->modelClass.$name", $callable);
+        $this->customRelations[$name] = $callback;
 
         return $this;
     }
 
     public function extend(Container $container, Extension $extension = null)
     {
-        // Nothing needed here.
+        foreach ($this->customRelations as $name => $callback) {
+            if (is_string($callback)) {
+                $callback = $container->make($callback);
```

Comment:
My concern here is that this gets resolved very early. This could create a NotificationMailer situation.

If we don't want to do the resolving in `AbstractModel`, we could do it with an anonymous callback, something like (untested):

```suggestion
                $callback = function () use ($container) {
                    return $container->call($callback, func_get_args());
                };
```

Reply:
I like this idea, but we might want to extract this out into a util so it can also be used for the `Formatter` and `Frontend` extensions. Maybe an `ExtenderUtil` class with static methods? Not 100% sure.

Reply:
Yes for code re-use, but calling the util would take almost as many lines of code as what it's replacing, no? :sweat_smile:

Reply:
It turns out to be a lot nicer actually. Only thing I'm unsure of is where to put the util (namespace + class name), I don't think we have too many of these right now?

Reply:
we could potentially put the class in `Flarum\Support` similar to where laravel puts its helper classes, that could be nice especially if we end up with more helper classes in the future.

But I personally don't mind either way.